### PR TITLE
Error when emitting the declaration for an exported class that extends an intersection

### LINKED ISSUE #14017: Compile errors with a generic base class and class mixins using declarative typings
**TypeScript Version:**  2.2.0 (rc) & 2.2.0-dev.20170211

I am having trouble creating consumable typings when using class mixins for a generic class from a core library - github example repro here https://github.com/agubler/typescript-mixin-typings.

Consider the following provided by a core library

```ts
// BaseClass.ts
export type Constructor<T> = new (...args: any[]) => T;

export class MyBaseClass<T> {
    baseProperty: string;

    constructor(value: T) {}
}
```

```ts
// MixinClass.ts
import { Constructor, MyBaseClass } from './BaseClass.ts';

export interface MyMixin {
    mixinProperty: string;
}

export function MyMixin<T extends Constructor<MyBaseClass<any>>>(base: T): T & Constructor<MyMixin> {
    mixinProperty: string;
}
```

```ts
// FinalClass
export class MyExtendedClass extends MyMixin(MyBaseClass)<string> {
	extendedClassProperty: number;
}
```

compiling with `declarations` `true`

```json
{
		"declaration": true,
		"emitDecoratorMetadata": true,
		"experimentalDecorators": true,
		"module": "umd",
		"moduleResolution": "node",
		"noImplicitAny": true,
		"noImplicitThis": true,
		"removeComments": false,
		"sourceMap": true,
		"strictNullChecks": true,
		"target": "es5"
}
```

Usage in downstream consumer project:

```ts
import { MyExtendedClass } from 'core-library/FinalClass';
import { MyMixin } from 'core-library/MixinClass';

// constructor signature lost
const myExtendedClass = new MyExtendedClass('string');

// MyExtendedClass no longer type MyBaseClass
const AnotherMixedClass = MyMixin(MyExtendedClass);
```

Errors:
`5 col 25 error| Supplied parameters do not match any signature of call target. [typescript/tsuquyomi]`

`8 col 35 error| Argument of type 'typeof MyExtendedClass' is not assignable to parameter of type 'Constructor<MyBaseClass<any>>'. Type 'MyExtendedClass' is not assignable to type 'MyBaseClass<any>'. Property 'baseProperty' is missing in type 'MyExtendedClass'. [typescript/tsuquyomi]`

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #14017 by making the mixin emit illegal.

### CODE REVIEW

**Comment by mhegazy** [comment]
@sandersn can you also file an issue to track parsing the intersection types in heritage clauses.

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -2352,6 +2352,10 @@
         "category": "Error",
         "code": 4092
     },
+    "Extends clause of exported class '{0}' refers to a type with no declaration.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: a type whose name cannot be referenced.

**Comment by sandersn** [comment]
@mhegazy Done: #14075

File: src/compiler/diagnosticMessages.json [hunk 2]
```diff
@@ -2352,6 +2352,10 @@
         "category": "Error",
         "code": 4092
     },
+    "Extends clause of exported class '{0}' refers to a type whose name cannot be referenced.": {
```
Comment (DanielRosenwasser) [inline src/compiler/diagnosticMessages.json]: Nit: `"The 'extends clause of exported class...`

Comment (sandersn) [inline src/compiler/diagnosticMessages.json]: Are you sure? All the other messages like this just start with "Extends clause ...".

Comment (sandersn) [inline src/compiler/diagnosticMessages.json]: A couple of minutes with @RyanCavanaugh turned up another option we liked even more: 'extends' clause of exported class ...

I'll go with that and update all the other error messages to match.

**Comment by mhegazy** [comment]
@sandersn please cherry-pick this in release-2.2

**Comment by sandersn** [comment]
@mhegazy done

**Pull request merged**